gslib/utils/copy_helper.py

```diff
@@ -2196,6 +2199,9 @@ def CallResumableUpload():
                               provider=dst_url.scheme)
       raise
 
+  if using_temporary_file:
+    os.unlink(src_obj_filestream.name)
```

Comment:
It might be nice if we catch this error and check if the error is NOT_EXISTS before re-raising.

Reply:
To make sure we're on the same page, in this upload scenario:
```
cp x.txt gs://my-bucket
```
The temporary file at `src_obj_filestream.name` is `x.txt_.stet_tmp`.

Deleting the file isn't related to an error. Are you referring to the possibility of an error during the unlink?

In that case, I think the fact the user added the `--stet` flag combined with the built-in not found error for a STET-extension file is sufficient:
```
>>> os.unlink('x.txt_.stet_tmp')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory: 'x.txt_.stet_tmp'
```

Reply:
I'm suggesting that, as an end user, I would have no idea what that error meant, nor would I know what I needed to do to fix it.

In the case where unlinking fails because the destination was already deleted, I think we should swallow that error and return success. In other situations, it would be helpful to add text like "you may need to delete this file manually".

Reply:
I can think of three scenarios where the temporary file would already be deleted:
1) The STET binary deletes its output file, but does not raise an error. In this case, gsutil continues execution to the `os.unlink`.
2) The user manually deletes the temporary file during a small window of time while cp is running.
3) Same as (2), but an automated program deletes the file.

(1) seems unlikely, and, if (2) occurred, the user would probably realize what happened based on their recent action and the built-in error message. I imagine a system policy blocking the creation of files is more likely than (3), in which case the error would happen for uploads within the STET binary.

***

For the second idea, it might be beneficial to add "you may need to delete x_.stet_tmp manually" to any errors the STET binary raises. However, I'm leaning away from it b/c I think it's clear from the binary raising an error message and the file extension of the temp file (".stet_tmp") what happened. Rerunning the command will also take care of the temp file.

Leaving this open in case other reviewers want to comment.

Reply:
There are many, many reasons why a temporary file would be deleted without user interaction, especially if the file is in a directory that the OS has flagged as tmp. These could include, but are not limited to, cron jobs, configuration management tools from corp IT, routine system cleanup operations, software like CCleaner/CleanMyMac, unplugging a USB drive...

I'm mostly just advocating for something more actionable to the end-user.

Reply:
In the latest design, gsutil adds the temporary files to the same directory as the original source or final destination files. Automated tools would have to distinguish between legitimate copy files and temp files within the same directory to cause this error. Gsutil uses other temporary files for resuming downloads, and users currently don't let their automated tools delete these.

Handling every scenario like a cron job or cleaner tool sweeping in to delete a temp file moments after its created would also make the code unwieldy. We would have to check for this in at least these places:
1) After the STET binary completes writing the file and before the upload begins.
2) After a download completes and before the STET binary begins decrypting it.
3) Before the STET binary tries to delete the original encrypted version of a file after decrypting it.

Referring to an earlier comment, in my opinion, the user gets enough hint about what happened from three facts:
1) The command was run with "--stet".
2) The error is related to a file ending in "stet_tmp" not being found.
3a) The user is aware the delete action they just took OR (3b)
3b) The user has knowledge about automatic deletions on their system.

Agreed that (3b) is the weakest point, but gsutil doesn't fall into the category of software that has to take an inventory of everything else happening on a user device and defend against unlikely scenarios.

Reply:
FYI, I'm OnDuty and inherited a backlog of 20+ bugs, so I don't really have time to discuss this further. If the majority of reviewers want to catch `FileNotFoundError` on delete operations, I'll do it.

Reply:
I feel we will have even bigger problems if a cronjob is deleting gsutil's tmp files. What if it deletes the tmp file while the upload is in progress? In fact, if such a cronjob exists, the probability of it deleting an in-progress file is much higher compared to the short window when os.unlink happens. So I feel the protection against os.unlink failing is not really needed here

I agree with @sethvargo  that the message to the user is not that great, but we make certain assumptions here that the user will not mess up with gsutil specific files to keep the code complexity low like Nick has mentioned. (We will have to add these checks at a lot of places if we really want to give the user a good message for situations like these.)

Reply:
Going to resolve to clean up the code review screen. Feel free to reopen.